src/io/rainrobot/maxdispachreq/AvailabilityService.java

```diff
@@ -0,0 +1,61 @@
+package io.rainrobot.maxdispachreq;
+
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Set;
+
+public class AvailabilityService {
+
+    private int maxAvailability;
+    private Map<Integer, Set<ServerAddress>> avelabiletyServerSetMap;
+    private AddressAvelabiletyTracker addressAvelabiletyTracker;
+
+    private void recomputeMaxAvailability() {
+        if (getMaxAvailabilitySet().size() == 0) {
+            do{
+                maxAvailability --;
+            } while (getMaxAvailabilitySet().size() == 0
+                                    && maxAvailability != 1);
+            maxAvailability --;
+        }
+    }
+    
+    private Set<ServerAddress> getMaxAvailabilitySet() {
+        return avelabiletyServerSetMap.get(maxAvailability);
+    }
+
+    public ServerAddress popFromMaxAvailableSet() {
+        ServerAddress popServer = getMaxAvailabilitySet().iterator().next();
```

Comment:
don't you always return the first here?

Reply:
Yes it's a mistake, the follow line should remove it from the set